[reopen] Rename `NavigationDestination` to `NavigationBarDestination` + some type hinting, fixes and new features

bleudev opened the pull request:
Checklist:

- [x] Navigation bar's `NavigationDestination` renamed to `NavigationBarDestination`.
- [x] Type hinting:
    - [x] Add type alias:
        - [x] Number
        - [x] ControlEventFunction
        - [x] IconStr
        - [x] ColorStr
        - [x] Wrapper
    - [x] `navigation_bar.py`, `navigation_drawer.py`, `navigation_rail.py`
    - [x] `cupertino_navigation_bar.py`
    - [x] `page.py`
    - [x] `control.py`
- [x] Move `OptionalNumber` to `types.py`
- [x] `page.py`:
    - [x] Add `close_snack_bar() method`
    - [x] #3173
    - [x] Add Exceptions when page controls is None in `Page.close_...()` methods 
    - [x] Rename `show_...()` to `open_...()`
- [x] Magic methods:
    - [x] `Control in Page`

- [x] `isort` and `black` formatting

(closes #3173)
(closes #3329)
(reopen of #3161)

It should be merged with [#281](https://github.com/flet-dev/website/pull/281)

ndonkoHenri commented on sdk/python/packages/flet-core/src/flet_core/control.py:
```diff
@@ -198,11 +197,13 @@ def page(self, page):
         self.__page = page
 
     # uid
+
```
Seems like your formatter is still on. 😅

bleudev replied:
I'll fix that

bleudev commented:
@ndonkoHenri May i merge your pull (#3053) into my pull (write your code to my pull). then it will be more convenient. Author of project will have to merge only my pull

bleudev commented:
@FeodorFitsner Can you review this PR?

ndonkoHenri commented:
> @ndonkoHenri Do i can merge your pull (#3053) into my pull (write your code to my pull)? then it will be more convenient. Author of project will have to merge only my pull

Let's keep everything separate. It eases review.

bleudev commented:
> > @ndonkoHenri Do i can merge your pull (#3053) into my pull (write your code to my pull)? then it will be more convenient. Author of project will have to merge only my pull
> 
> Let's keep everything separate. It eases review.

I have already implemented this :)

ndonkoHenri commented:
> I have already implemented this

Alright, didn't see that before.
Please next time let's keep things separate when possible :) 

Furthermore, Autoupdate is actually still in discussion, reason why I didn't wrap it up.

bleudev commented:
> > I have already implemented this
> 
> Alright, didn't see that before. Please next time let's keep things separate when possible :)
> 
> Furthermore, Autoupdate is actually still in discussion, reason why I didn't wrap it up.

When my pull will be merged, your features will also be merged. Can you review my pull for to make it happen faster

FeodorFitsner commented:
@bleudev, @ndonkoHenri So this PR is based on a branch from https://github.com/flet-dev/flet/pull/3053?

bleudev commented:
> @bleudev, @ndonkoHenri So this PR is based on a branch from #3053?

No, i just implemented features from henri's pull and added my features

bleudev commented:
> > @bleudev, @ndonkoHenri So this PR is based on a branch from #3053?
> 
> No, i just implemented features from henri's pull and added my features

More accurately, i just use he's idea, but code is mine

bleudev commented:
@FeodorFitsner

bleudev commented:
@FeodorFitsner @ndonkoHenri Can you review my PR?

FeodorFitsner commented:
I'm going to look into this, and other PRs early next week when I'm back from PyCon. Thanks for your efforts and sorry for the delay with feedback.

bleudev commented:
> I'm going to look into this, and other PRs early next week when I'm back from PyCon. Thanks for your efforts and sorry for the delay with feedback.

Ok, i'll do my PR on website repo for get ready to merge.

FeodorFitsner commented on sdk/python/packages/flet-core/src/flet_core/column.py:
```diff
@@ -169,6 +171,15 @@ def clean(self):
+    def remove(self, *controls: Control) -> None:
+        self.__remove(self.__controls, *controls)
+
     # tight
```
Could we remove all occurrences of `add`, `insert` and `remove` (and underlying logic) from the PR please?
1) We would have auto-update related feature in a separate PR and 2) I've been recently thinking a lot about auto-update feature, which is of course will be great (and maybe "a must") addition to Flet, but came out with a different approach. Those `add/insert/remove` was added in "experimental" PR by @ndonkoHenri and, perhaps, won't be needed at all in a new approach.

I think we could close https://github.com/flet-dev/flet/pull/3053 as well.

Do we have an issue to discuss auto-update?

bleudev replied:
> Could we remove all occurrences of `add`, `insert` and `remove` (and underlying logic) from the PR please?
> 
> 1. We would have auto-update related feature in a separate PR and 2) I've been recently thinking a lot about auto-update feature, which is of course will be great (and maybe "a must") addition to Flet, but came out with a different approach. Those `add/insert/remove` was added in "experimental" PR by @ndonkoHenri and, perhaps, won't be needed at all in a new approach.
> 
> I think we could close #3053 as well.
> 
> Do we have an issue to discuss auto-update?

I want to implement this feature now because it's logical (Page have `add()`, `insert()` and `remove()`. If i have to remove this methods, i have to remove `__add__()`, `__radd__()`, `__iadd__()` and other magic methods. Also i already did `add/insert/remove` methods in [#281](https://github.com/flet-dev/website/pull/281). I don't want to re-do this and that PR.

Why i and other contributors can't do `add/insert/remove` methods?

FeodorFitsner replied:
I appreciate your time working on this PR, but auto-update feature won't be working like "change collection, update control/page" - thus those methods are unnecessary. Without `self.update` these methods don't add any value (`self.controls.append(something)` vs `self.add(something)`) and just make API more complex.

bleudev replied:
> I appreciate your time working on this PR, but auto-update feature won't be working like "change collection, update control/page" - thus those methods are unnecessary. Without `self.update` these methods don't add any value (`self.controls.append(something)` vs `self.add(something)`) and just make API more complex.

Thank you for feedback, but for me these methods is very comfortable. It's possible to make these methods and not making API more complex? Also how to make these methods less unnecessary?

FeodorFitsner replied:
Let's try avoiding to do mega-PRs including multiple features (I do my best to hold myself putting everything into one PR as well :) If you are able to break your PR(s) into smaller PRs then we could discuss them separately.

I'm not saying these methods are not usable or unnecessary I just don't want to include them into core API now to keep things simple. Not everything could be included into core. Hope you understand. If you need those methods you can create a separate package "patching" Flet classes and put it on pypi.org.

For better communication I'm going to setup a public channel for Flet developers contributing or willing to contribute to Flet core. There we could discuss whether this or that functionality is desirable, how to implement this. Also, we are going to have periodic video hang-outs there.

bleudev replied:
> Let's try avoiding to make mega-PRs including multiple features (I do my best to hold myself putting everything into one PR as well :) If you are able to break your PR(s) into smaller PRs then we could discuss them separately. It's easier for developers and maintainers.
> 
> I'm not saying these methods are not usable or unnecessary I just don't want to include them into core API now to keep things simple. Not everything could be included into core. Hope you understand. If you need those methods you can create a separate package "patching" Flet classes and put it on pypi.org.
> 
> For better communication I'm going to setup a public channel for Flet developers contributing or willing to contribute to Flet core. There we could discuss whether this or that functionality is desirable, how to implement this. Also, we are going to have periodic video hang-outs there.

Ok, i'll remove this methods

ndonkoHenri replied:
#3329

FeodorFitsner commented on sdk/python/packages/flet-core/src/flet_core/types.py:
```diff
@@ -290,3 +297,20 @@ class MouseCursor(Enum):
+
+
+# Events
+ControlEventFunction = Callable[[ControlEvent], None]
```
I think this should be `Optional`:

![image](https://github.com/flet-dev/flet/assets/5041459/61f7a051-a8fe-4a04-bd92-931d2349ab98)

bleudev replied:
Ok, thanks. I'll fix that

FeodorFitsner commented on sdk/python/packages/flet-core/src/flet_core/types.py:
```diff
@@ -73,7 +74,13 @@ class NotchShape(Enum):
+# str type alias
+ColorStr = Optional[str]
+IconStr = Optional[str]
 
```
Love the idea of `ColorStr` and `IconStr`! Would it be possible/make sense to put that in *all* controls?

FeodorFitsner replied:
Alternatively, we could convert `icons.py` and `colors.py` to Enums and use them for even better typing. Could be a separate issue/PR though.

ndonkoHenri replied:
Yep, I think Enum is the way to go.

bleudev replied:
> Alternatively, we could convert `icons.py` and `colors.py` to Enums and use them for even better typing. Could be a separate issue/PR though.

I can make in PR after merging

FeodorFitsner commented on sdk/python/packages/flet-core/src/flet_core/view.py:
```diff
@@ -275,3 +275,7 @@ def fullscreen_dialog(self) -> Optional[bool]:
+
+    # Magic methods
+    def __contains__(self, item: Control) -> bool:
+        return item in self.__controls
```
Could you please elaborate what's the purpose of this method?

FeodorFitsner commented on sdk/python/packages/flet-core/src/flet_core/page.py:
```diff
@@ -149,6 +171,7 @@ def __init__(
         self.__loop = loop
         self.__executor = executor
         self._index = {self._Control__uid: self}  # index with all page controls
+        self.__auto_update = False
```
Can be removed.

bleudev replied:
I don't understand why

FeodorFitsner replied:
It's not used now. Why keep it?

FeodorFitsner commented on sdk/python/packages/flet-core/src/flet_core/page.py:
```diff
@@ -974,42 +1001,112 @@ def __on_invoke_method_result(self, e):
-        self.__offstage.update()
+    @overload
+    def open_snack_bar(self) -> None:
+        ...
```
Why do we need guys like these?

FeodorFitsner replied:
Btw, I was thinking about unifying of all "openable" controls and having a single `open()` method for everything. Like adding `OpenableControl` (what would be a better name? :) base class with at least `open` property and then having a list of such controls in `__offstage`. Does it make sense?

bleudev replied:
I can open PR after merging this with realisation of open() method

FeodorFitsner commented:
FYI: Created a new channel for devs: https://discord.com/channels/981374556059086931/1243237230080688199

FeodorFitsner commented on sdk/python/packages/flet-core/src/flet_core/types.py:
```diff
@@ -290,3 +297,20 @@ class MouseCursor(Enum):
+# Protocols
+class SupportsStr(Protocol):
+    def __str__(self) -> str:
+        ...
```
What is this for?

bleudev replied:
This is protocol for type hinting. SupportsStr is an object that supports str()

bleudev commented:
> Great PR with valuable typings additions - much appreciated!

Thanks

bleudev commented on sdk/python/packages/flet-core/src/flet_core/page.py:
```diff
@@ -1175,57 +1346,68 @@ def window_close(self):
+    @auto_update.setter
+    def auto_update(self, value: bool):
+        self.__auto_update = value
+
```
i don't understand why

bleudev replied:
> Our ultimate goal is to get rid of `.update()` method completely. I've seen a lot of complaints on the internet that "it requires to call .update() every time...don't forget that! etc.". Auto-update will be on by default. .update() will stay there and could be treated as analogue of "flush()" in streams.

I still don't get it why this `can be removed`. Sorry for my not understanding

FeodorFitsner replied:
OK, "please remove it" to being more straightforward ;)

bleudev replied:
> OK, "please remove it" to being more straightforward ;)

Remove property? On this way i have to rename `self.__auto_update` to `self.auto_update`.

FeodorFitsner replied:
Both property and self.__auto_update as they're no longer needed.

bleudev replied:
> Both property and self.__auto_update as they're no longer needed.

Why? You request to delete `auto_update` feature? Or use `auto_update` as always `True`?

FeodorFitsner replied:
Right, I request to delete auto_update feature

bleudev replied:
> Right, I request to delete auto_update feature

Why? It's very useful function and simple.

bleudev replied:
@FeodorFitsner Why you request to delete auto_update?

FeodorFitsner replied:
As I explained before...This entire "auto-update" thing is going to work differently. Calling .update() after every property change is an overhead. I don't want to rush with that feature yet and merge it half-baked. This PR has a lot of good stuff - I really don't want to close it as I appreciate the time you spent already.

bleudev replied:
> As I explained before...This entire "auto-update" thing is going to work differently. Calling .update() after every property change is an overhead. I don't want to rush with that feature yet and merge it half-baked. This PR has a lot of good stuff - I really don't want to close it as I appreciate the time you spent already.

ok, i'll delete auto_update feature

ndonkoHenri commented:
https://github.com/flet-dev/flet/issues/3348
I personally think we should leave colors and icons as-is. I mean, what is honestly the issue with `Optional[str]` ? it's short and normal - nothing fancy, no? 
However, if you dont think its the case, then I suggest you could create something like `OptionalString=Optional[str]`, then use it where necessary. I dont see the need of having `HexStr`, `ColorStr`, `IconStr`, etc, which are at the end of the day thesame (`= Optional[str]`)

bleudev commented:
@FeodorFitsner re-review please)

FeodorFitsner commented:
Reviewing it now. Thanks!

bleudev commented:
Thank you very much!
